File: android/src/main/java/com/margelo/nitro/svga/ProtoReader.kt
package com.margelo.nitro.svga

internal class ProtoReader private constructor(
  private val buf: ByteArray,
  private val limit: Int,
  private var pos: Int
) {

  companion object {
    const val WIRE_VARINT = 0
    const val WIRE_FIXED64 = 1
    const val WIRE_LENGTH_DELIMITED = 2
    const val WIRE_FIXED32 = 5

    private const val MAX_VARINT_BYTES = 10
  }

  constructor(bytes: ByteArray) : this(bytes, bytes.size, 0)

  data class Tag(val field: Int, val wire: Int)

  fun hasMore(): Boolean = pos < limit

  fun readVarint(): Long {
    var result = 0L
    var shift = 0
    var read = 0
    while (pos < limit) {
      if (read >= MAX_VARINT_BYTES) throw SvgaParseException("varint too long")
      val b = buf[pos].toLong() and 0xFF
      pos++
      read++
      result = result or ((b and 0x7F) shl shift)
      if ((b and 0x80) == 0L) return result
      shift += 7
    }
    throw SvgaParseException("truncated varint")
  }

  fun readTag(): Tag {
    val raw = readVarint().toInt()
    return Tag(raw ushr 3, raw and 0x7)
  }

  fun readSubReader(): ProtoReader {
    val len = checkedLength()
    val sub = ProtoReader(buf, pos + len, pos)
    pos += len
    return sub
  }

  fun readBytes(): ByteArray {
    val len = checkedLength()
    val data = buf.copyOfRange(pos, pos + len)
    pos += len
    return data
  }

  fun readString(): String {
    val len = checkedLength()
    val s = String(buf, pos, len, Charsets.UTF_8)
    pos += len
    return s
  }

  fun readInt32(): Int = readVarint().toInt()

  fun readFloat(): Float {
    if (pos + 4 > limit) throw SvgaParseException("truncated float")
    val bits = (buf[pos].toInt() and 0xFF) or
      ((buf[pos + 1].toInt() and 0xFF) shl 8) or
      ((buf[pos + 2].toInt() and 0xFF) shl 16) or
      ((buf[pos + 3].toInt() and 0xFF) shl 24)
    pos += 4
    return java.lang.Float.intBitsToFloat(bits)
  }

  fun skip(wire: Int) {
    when (wire) {
      WIRE_VARINT -> readVarint()
      WIRE_FIXED64 -> advance(8)
      WIRE_LENGTH_DELIMITED -> skipBytes()
      WIRE_FIXED32 -> advance(4)
      else -> throw SvgaParseException("unknown wire type $wire")
    }
  }

  private fun skipBytes() {
    val len = checkedLength()
    pos += len
  }

  private fun advance(n: Int) {
    if (pos + n > limit) throw SvgaParseException("truncated fixed field")
    pos += n
  }

  /// Read a length-delimited length and validate it lies within the reader's
  /// window. Performed in Long arithmetic so a hostile varint near
  /// `Int.MAX_VALUE` cannot wrap `pos + len` to a small positive number and
  /// slip the bounds check, which would then trap with
  /// NegativeArraySizeException / StringIndexOutOfBoundsException further on.
  private fun checkedLength(): Int {
    val raw = readVarint()
    if (raw < 0 || raw > Int.MAX_VALUE.toLong()) {
      throw SvgaParseException("length-delimited overflow")
    }
    val len = raw.toInt()
    val end = pos.toLong() + len.toLong()
    if (end > limit.toLong()) throw SvgaParseException("length-delimited overflow")
    return len
  }
}

internal class SvgaParseException(message: String) : RuntimeException(message)

File: ios/ProtoReader.swift
import Foundation

internal final class ProtoReader {
    enum Wire: Int {
        case varint = 0
        case fixed64 = 1
        case lengthDelimited = 2
        case fixed32 = 5
    }

    struct Tag {
        let field: Int
        let wire: Int
    }

    private let data: Data
    private let base: Int
    private let limit: Int
    private var pos: Int

    init(_ data: Data) {
        self.data = data
        self.base = data.startIndex
        self.limit = data.endIndex
        self.pos = data.startIndex
    }

    private init(data: Data, base: Int, limit: Int) {
        self.data = data
        self.base = base
        self.limit = limit
        self.pos = base
    }

    var hasMore: Bool { pos < limit }

    func readVarint() throws -> UInt64 {
        var result: UInt64 = 0
        var shift: UInt64 = 0
        var read = 0
        while pos < limit {
            if read >= 10 { throw SvgaError("varint too long") }
            let b = UInt64(data[pos])
            pos += 1
            read += 1
            result |= (b & 0x7F) << shift
            if (b & 0x80) == 0 { return result }
            shift += 7
        }
        throw SvgaError("truncated varint")
    }

    func readTag() throws -> Tag {
        let raw = Int(try readVarint())
        return Tag(field: raw >> 3, wire: raw & 0x7)
    }

    func readSubReader() throws -> ProtoReader {
        let len = try checkedLength()
        let sub = ProtoReader(data: data, base: pos, limit: pos + len)
        pos += len
        return sub
    }

    func readBytes() throws -> Data {
        let len = try checkedLength()
        let slice = data.subdata(in: pos..<(pos + len))
        pos += len
        return slice
    }

    func readString() throws -> String {
        let len = try checkedLength()
        // subdata copies; required because the caller may outlive `data` and
        // a slice would retain the whole buffer.
        let slice = data.subdata(in: pos..<(pos + len))
        pos += len
        return String(decoding: slice, as: UTF8.self)
    }

    func readInt32() throws -> Int32 { Int32(truncatingIfNeeded: try readVarint()) }

    func readFloat() throws -> Float {
        if pos + 4 > limit { throw SvgaError("truncated float") }
        let bits = UInt32(data[pos])
            | (UInt32(data[pos + 1]) << 8)
            | (UInt32(data[pos + 2]) << 16)
            | (UInt32(data[pos + 3]) << 24)
        pos += 4
        return Float(bitPattern: bits)
    }

    func skip(wire: Int) throws {
        switch wire {
        case Wire.varint.rawValue: _ = try readVarint()
        case Wire.fixed64.rawValue: try advance(8)
        case Wire.lengthDelimited.rawValue: try skipBytes()
        case Wire.fixed32.rawValue: try advance(4)
        default: throw SvgaError("unknown wire type \(wire)")
        }
    }

    private func skipBytes() throws {
        let len = try checkedLength()
        pos += len
    }

    private func advance(_ n: Int) throws {
        if pos + n > limit { throw SvgaError("truncated fixed field") }
        pos += n
    }

    /// Reads a length-delimited length and validates it lies within the
    /// reader's window. Performed in Int64 space so a hostile varint near
    /// `Int.max` cannot wrap `pos + len` to a small positive number and slip
    /// the bounds check.
    private func checkedLength() throws -> Int {
        let raw = try readVarint()
        if raw > UInt64(Int.max) { throw SvgaError("length-delimited overflow") }
        let len = Int(raw)
        let end = Int64(pos) + Int64(len)
        if len < 0 || end > Int64(limit) {
            throw SvgaError("length-delimited overflow")
        }
        return len
    }
}
